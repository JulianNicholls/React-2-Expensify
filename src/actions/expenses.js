import database from '../firebase';

import { ADD_EXPENSE, REMOVE_EXPENSE, EDIT_EXPENSE, SET_EXPENSES } from './types';

// ADD EXPENSE

export const startAddExpense = (expenseData = {}) => {
  return dispatch => {
    const { description = '', note = '', amount = 0, createdAt = 0 } = expenseData;

    const expense = { description, note, amount, createdAt };

    return database
      .ref('expenses')
      .push(expense)
      .then(ref => dispatch(addExpense({ id: ref.key, ...expense })));
  };
};

export const addExpense = expense => ({ type: ADD_EXPENSE, expense });

// REMOVE EXPENSE

export const startRemoveExpense = id => {
  return dispatch => {
    return database
      .ref(`expenses/${id}`)
      .remove()
      .then(() => dispatch(removeExpense(id)));
  };
};

export const removeExpense = id => ({ type: REMOVE_EXPENSE, id });

// EDIT EXPENSE

export const startEditExpense = (id, updates) => {
  return dispatch => {
    return database
      .ref(`expenses/${id}`)
      .update(updates)
      .then(() => dispatch(editExpense(id, updates)));
  };
};

export const editExpense = (id, updates) => ({ type: EDIT_EXPENSE, id, updates });

// SET EXPENSES

export const startSetExpenses = () => {
  return dispatch => {
    return database
      .ref('expenses')
      .once('value')
      .then(snapshot => {
        const expensesData = [];

        snapshot.forEach(item => {
          expensesData.push({ id: item.key, ...item.val() });
        });

        dispatch(setExpenses(expensesData));
      });
  };
};

export const setExpenses = expenses => ({ type: SET_EXPENSES, expenses });
