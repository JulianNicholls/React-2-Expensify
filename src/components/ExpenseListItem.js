import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
import 'numeral/locales/en-gb';

numeral.locale('en-gb');

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <div style={{ display: 'flex', alignItems: 'center', marginBottom: '7px' }}>
    <Link style={{ width: '60%' }} to={`/edit/${id}`}>
      <h4 style={{ margin: '0' }}>{description}</h4>
    </Link>
    <div style={{ width: '15%', textAlign: 'right' }}>
      {numeral(amount / 100.0).format('$0,0.00')}
    </div>
    <div style={{ width: '15%', paddingLeft: '2em' }}>
      {moment(createdAt).format('ll')}
    </div>
  </div>
);

export default ExpenseListItem;
