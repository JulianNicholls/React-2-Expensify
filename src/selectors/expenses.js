export default (expenses, { text, sortBy, startDate, endDate }) => {
  text = text.toLowerCase();

  const filtered = expenses.filter(({ description, createdAt }) => {
    const startDateMatch = startDate ? startDate <= createdAt : true;
    const endDateMatch = endDate ? endDate >= createdAt : true;
    const textMatch = text === '' || description.toLowerCase().includes(text);

    return startDateMatch && endDateMatch && textMatch;
  });

  // Sort the expenses, both date and amount are sorted into descending order.

  return filtered.sort((a, b) => {
    if (sortBy === 'date') return b.createdAt - a.createdAt;

    return b.amount - a.amount;
  });
};
