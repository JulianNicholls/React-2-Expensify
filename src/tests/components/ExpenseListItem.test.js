import React from 'react';
import { shallow } from 'enzyme';

import ExpenseListItem from '../../components/ExpenseListItem';
import { testExpenses } from '../fixtures/expenses';

describe('ExpenseListItem component', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<ExpenseListItem {...testExpenses[1]} />);

    expect(wrapper).toMatchSnapshot();
  });
});
