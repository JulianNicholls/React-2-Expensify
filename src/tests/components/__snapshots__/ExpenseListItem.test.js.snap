// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseListItem component should render correctly 1`] = `
<Link
  className="list-item"
  replace={false}
  to="/edit/2"
>
  <div>
    <h3
      className="list-item__title"
    >
      rates
    </h3>
    <span
      className="list-item__subtitle"
    >
      Jan 21, 1970
    </span>
  </div>
  <h3
    className="list-item__data"
  >
    Â£150.00
  </h3>
</Link>
`;
